File: office-portal/apps/monthly_tracking/services/hesaplama.py
"""Aylik ozet hesaplama — aylik-takip/hesaplama.js'in birebir Python portu."""
from __future__ import annotations

import calendar

from .resmi_tatiller import ayin_resmi_tatilleri

GUNLUK_SAAT = 9


def ayin_is_gunu_sayisi(yil: int, ay: int) -> int:
    gun_sayisi = calendar.monthrange(yil, ay)[1]
    return sum(1 for g in range(1, gun_sayisi + 1) if _tarih_haftaici(yil, ay, g))


def _tarih_haftaici(yil, ay, gun):
    import datetime

    return datetime.date(yil, ay, gun).weekday() < 5  # Pazartesi=0..Cuma=4


def personel_ozet(*, isim, calisma_saat, izin_gun, is_gunu, resmi_gun):
    # izin_gun artik yarim gun/saatlik izinler yuzunden ondalikli olabilir
    # (orn. 4.5) — goruntulemede gereksiz uzun ondalik kalmasin diye yuvarlanir.
    izin_gun = round(izin_gun * 100) / 100
    izin_saat = izin_gun * GUNLUK_SAAT
    resmi_saat = resmi_gun * GUNLUK_SAAT

    calisma_gerekli_gun = max(0, is_gunu - resmi_gun)
    gerekli_saat = calisma_gerekli_gun * GUNLUK_SAAT

    fark = round((calisma_saat + izin_saat - gerekli_saat) * 100) / 100

    return {
        "isim": isim,
        "calisma_saat": round(calisma_saat * 100) / 100,
        "izin_gun": izin_gun,
        "izin_saat": izin_saat,
        "resmi_gun": resmi_gun,
        "resmi_saat": resmi_saat,
        "calisma_gerekli_gun": calisma_gerekli_gun,
        "gerekli_saat": round(gerekli_saat * 100) / 100,
        "fark": fark,
        "durum": "tam" if abs(fark) < 0.01 else ("fazla" if fark > 0 else "eksik"),
    }


def aylik_ozet(*, personeller, izinler, isim_eslesir, yil, ay):
    is_gunu = ayin_is_gunu_sayisi(yil, ay)
    resmi_tatiller = ayin_resmi_tatilleri(yil, ay)
    resmi_hafta_ici = [t for t in resmi_tatiller if t["hafta_ici"]]
    resmi_gun = len(resmi_hafta_ici)

    satirlar = []
    for per in personeller:
        kendi_izinleri = [iz for iz in izinler if isim_eslesir(per["isim"], iz["isim"])]
        izin_gun = sum(iz.get("is_gunu", 0) for iz in kendi_izinleri)
        ozet = personel_ozet(
            isim=per["isim"], calisma_saat=per["toplam_saat"],
            izin_gun=izin_gun, is_gunu=is_gunu, resmi_gun=resmi_gun,
        )
        ozet.update({
            "sira": per.get("sira"),
            "bolgede_saat": per.get("bolgede_saat"),
            "muafiyet_saat": per.get("muafiyet_saat"),
            "izin_kayitlari": kendi_izinleri,
        })
        satirlar.append(ozet)

    eslesmeyen_izinler = [
        iz for iz in izinler if not any(isim_eslesir(per["isim"], iz["isim"]) for per in personeller)
    ]

    return {
        "yil": yil, "ay": ay, "gunluk_saat": GUNLUK_SAAT,
        "ayin_is_gunu": is_gunu, "resmi_gun": resmi_gun, "resmi_saat": resmi_gun * GUNLUK_SAAT,
        "resmi_tatiller": resmi_hafta_ici, "resmi_tatiller_tum": resmi_tatiller,
        "personeller": satirlar, "eslesmeyen_izinler": eslesmeyen_izinler,
    }

File: aylik-takip/hesaplama.js
// Aylik calisma/izin ozetini hesaplayan is mantigi.

const { ayinResmiTatilleri } = require('./resmiTatiller');

const GUNLUK_SAAT = 9; // gunluk mesai 9 saat kabul edilir

/**
 * Verilen yil/ay icin hafta ici (Pzt-Cuma) is gunu sayisini bulur.
 * @param {number} yil
 * @param {number} ay 1-12
 * @returns {number}
 */
function ayinIsGunuSayisi(yil, ay) {
  const gunSayisi = new Date(yil, ay, 0).getDate();
  let is = 0;
  for (let g = 1; g <= gunSayisi; g++) {
    const gun = new Date(yil, ay - 1, g).getDay(); // 0=Pazar,6=Cumartesi
    if (gun !== 0 && gun !== 6) is++;
  }
  return is;
}

/**
 * Tek bir personel icin aylik ozet hesaplar.
 *
 * Calismasi gereken saat = (ayin is gunu - resmi tatil) x 9
 * Izin suresi calismasi gereken saate dahildir (dusulmez); fark hesaplanirken
 * izinli gecirilen saatler fiili calisma saatine eklenir.
 *
 * @param {Object} p
 * @param {string} p.isim
 * @param {number} p.calismaSaat  PDF Toplam Sure (fiili calisilan saat)
 * @param {number} p.izinGun      Gorselden okunan izin gun sayisi (hafta ici)
 * @param {number} p.isGunu       Ayin toplam is gunu (Pzt-Cuma)
 * @param {number} p.resmiGun     Aydaki hafta ici resmi tatil gunu
 * @returns {Object}
 */
function personelOzet({ isim, calismaSaat, izinGun, isGunu, resmiGun }) {
  const izinSaat = izinGun * GUNLUK_SAAT;
  const resmiSaat = resmiGun * GUNLUK_SAAT;

  // Calismasi gereken gun = is gunu - resmi tatil (izin dusulmez).
  const calismaGerekliGun = Math.max(0, isGunu - resmiGun);
  const gerekliSaat = calismaGerekliGun * GUNLUK_SAAT;

  // Izinli gecirilen saatler fiili calisilmis gibi sayilir.
  const fark = Math.round((calismaSaat + izinSaat - gerekliSaat) * 100) / 100;

  return {
    isim,
    calismaSaat: Math.round(calismaSaat * 100) / 100,
    izinGun,
    izinSaat,
    resmiGun,
    resmiSaat,
    calismaGerekliGun,
    gerekliSaat: Math.round(gerekliSaat * 100) / 100,
    fark,
    durum: Math.abs(fark) < 0.01 ? 'tam' : fark > 0 ? 'fazla' : 'eksik',
  };
}

/**
 * Tum personel + izin kayitlarindan aylik ozet tablosu uretir.
 *
 * @param {Object} p
 * @param {Array<{isim:string,toplamSaat:number}>} p.personeller  PDF'ten
 * @param {Array<{isim:string,isGunu:number}>} p.izinler           Gorselden
 * @param {Function} p.isimEslesir  (a,b)=>bool isim karsilastirici
 * @param {number} p.yil
 * @param {number} p.ay 1-12
 * @returns {Object}
 */
function aylikOzet({ personeller, izinler, isimEslesir, yil, ay }) {
  const isGunu = ayinIsGunuSayisi(yil, ay);
  const resmiTatiller = ayinResmiTatilleri(yil, ay);
  const resmiHaftaIci = resmiTatiller.filter((t) => t.haftaIci);
  const resmiGun = resmiHaftaIci.length;

  const satirlar = personeller.map((per) => {
    // Bu personele ait izin kayitlarini isimle eslesir topla.
    const kendiIzinleri = izinler.filter((iz) => isimEslesir(per.isim, iz.isim));
    const izinGun = kendiIzinleri.reduce((t, iz) => t + (iz.isGunu || 0), 0);
    return {
      ...personelOzet({
        isim: per.isim,
        calismaSaat: per.toplamSaat,
        izinGun,
        isGunu,
        resmiGun,
      }),
      // PDF'ten gelen ek detaylar (saat kirilimlari).
      sira: per.sira,
      bolgedeSaat: per.bolgedeSaat,
      muafiyetSaat: per.muafiyetSaat,
      // Gorselden gelen izin kayitlari (tarih araligi + tip dahil).
      izinKayitlari: kendiIzinleri,
    };
  });

  // Izin kaydi olup PDF'te personeli bulunamayanlar (uyari icin).
  const eslesmeyenIzinler = izinler.filter(
    (iz) => !personeller.some((per) => isimEslesir(per.isim, iz.isim))
  );

  return {
    yil,
    ay,
    gunlukSaat: GUNLUK_SAAT,
    ayinIsGunu: isGunu,
    resmiGun,
    resmiSaat: resmiGun * GUNLUK_SAAT,
    resmiTatiller: resmiHaftaIci,
    resmiTatillerTum: resmiTatiller,
    personeller: satirlar,
    eslesmeyenIzinler,
  };
}

module.exports = { aylikOzet, personelOzet, ayinIsGunuSayisi, GUNLUK_SAAT };
